rdr_service/resource/calculators/participant_enrollment_status.py

```diff
@@ -0,0 +1,303 @@
+    calculated = False  # False = No, True = Yes
+    first_ts = None  # First timestamp seen.
+    last_ts = None  # Last timestamp seen.
+    values: list = list()  # List of related events.
+
+
+class EnrollmentStatusCalculator:
+    """
+    Calculate participant enrollment status.
+    """
+    status: PDREnrollmentStatusEnum = PDREnrollmentStatusEnum.Unset
+    # events = None  # List of EnrollmentStatusEvent objects.
+    activity = None  # List of activity created from Participant generator.
+
+    cohort = None
+
+    # First info object for each part used in calculating enrollment status.
+    _signup = None
+    _consented = None
+    _ehr_consented = None
+    _gror_consented = None
+    _biobank_samples = None
+    _physical_measurements = None
+    _modules = None
+
+    # Timestamps for when each status was achieved, these are set in the self.save_calc() method.
+    registered_ts = None
+    participant_ts = None
+    participant_plus_ehr_ts = None
+    core_participant_minus_pm_ts = None
+    core_participant_ts = None
+
+    def run(self, activity: list):
+        """
+        :param activity: A list of activity dictionary objects created by the ParticipantSummaryGenerator.
+        """
+        self.activity = sorted([JSONObject(r) for r in activity if r['timestamp']], key=lambda i: i.timestamp)
+        # Work through activity by slicing to determine current enrollment status.
+        # This method allows us to iterate once through the data and still catch participants
+        # that might have been considered a Core Participant at one point, but would not by
+        # looking at their current state.
+        for x in range(1, len(self.activity)+1):
+            events = self.activity[0:x]
+
+            # Get each datum needed for calculating the enrollment status.
+            signed_up = self.calc_signup(events)
+            consented, cohort = self.calc_consent(events)
+            ehr_consented = self.calc_ehr_consent(events)
+            gror_consented = self.calc_gror_consent(events)
+            biobank_samples = self.calc_biobank_samples(events)
+            physical_measurements = self.calc_physical_measurements(events)
+            modules = self.calc_modules(events)
+
+            if not self.cohort:
+                self.cohort = cohort
+
+            # Calculate enrollment status
+            status = PDREnrollmentStatusEnum.Unset
+            if signed_up:
+                status = PDREnrollmentStatusEnum.Registered
+            if status == PDREnrollmentStatusEnum.Registered and consented:
+                status = PDREnrollmentStatusEnum.Participant
+            if status == PDREnrollmentStatusEnum.Participant and ehr_consented:
+                status = PDREnrollmentStatusEnum.ParticipantPlusEHR
+            if status == PDREnrollmentStatusEnum.ParticipantPlusEHR and biobank_samples and \
+                    (modules and len(modules.values) >= 3) and \
```

Comment:
Done, see new commit.

Reply:
Sorry, I should have been clearer. What I meant with the comment here was that the config has the list of baseline modules expected, and right now there are 3 of them but that could change in the future. Rather than having 3 hardcoded here, can the number of baseline modules be loaded from the config?

Reply:
The new commit does that.  It loads the baseline modules from the cofing and creates a list of PDREnrollmentStatusEnum enums from it.  So if a new baseline module is added to the config it automatically gets added to the enrollment status calculation without requiring any code changes.

Reply:
So the `>= 3` here wouldn't need to change if a 4th baseline module was added and needed to be completed for a participant to be considered CORE?

Reply:
I missed the check with the hardcode '3', that is fixed now.